File: ts/earthbucks-lib/test/tx-signer.test.ts
import { describe, expect, test, beforeEach, it } from "vitest";
import { TxBuilder } from "../src/tx-builder.js";
import { TxOutBnMap } from "../src/tx-out-bn-map.js";
import { TxOut } from "../src/tx-out.js";
import { Script } from "../src/script.js";
import { KeyPair } from "../src/key-pair.js";
import { Pkh } from "../src/pkh.js";
import { PkhKeyMap } from "../src/pkh-key-map.js";
import { TxSigner } from "../src/tx-signer.js";
import { ScriptInterpreter } from "../src/script-interpreter.js";
import { HashCache } from "../src/tx.js";
import { FixedBuf } from "@webbuf/fixedbuf";
import type { WebBuf } from "@webbuf/webbuf";
import { TxOutBn } from "../src/tx-out-bn.js";
import { U8, U16BE, U32BE, U64BE } from "@webbuf/numbers";
import type { TxIn } from "../src/tx-in.js";

describe("TxSigner", () => {
  let txBuilder: TxBuilder;
  let txSigner: TxSigner;
  let txOutBnMap: TxOutBnMap;
  let pkhKeyMap: PkhKeyMap;

  beforeEach(() => {
    txOutBnMap = new TxOutBnMap();
    pkhKeyMap = new PkhKeyMap();
    // generate 5 keys, 5 outputs, and add them to the txOutMap
    for (let i = 0; i < 5; i++) {
      const key = KeyPair.fromRandom();
      const pkh = Pkh.fromPubKeyBuf(key.pubKey.toBuf());
      pkhKeyMap.add(key, pkh);
      const script = Script.fromPkhOutput(pkh);
      const txOut = new TxOut(new U64BE(100), script);
      const txOutBn = new TxOutBn(txOut, new U32BE(0n));
      txOutBnMap.add(txOutBn, FixedBuf.alloc(32), new U32BE(i));
    }

    const changeScript = Script.fromEmpty();
    txBuilder = new TxBuilder(txOutBnMap, changeScript, new U32BE(0n));
  });

  test("should sign a tx", () => {
    const key = KeyPair.fromRandom();
    const pkh = Pkh.fromPubKeyBuf(key.pubKey.toBuf());
    const script = Script.fromPkhOutput(pkh);
    const output = new TxOut(new U64BE(50), script);
    txBuilder.addOutput(output);

    const tx = txBuilder.build();

    expect(tx.inputs.length).toBe(1);
    expect(tx.outputs.length).toBe(2);
    expect(tx.outputs[0]?.value.bn).toEqual(BigInt(50));

    txSigner = new TxSigner(tx, txOutBnMap, pkhKeyMap, new U32BE(0n));
    const signed = txSigner.sign(new U32BE(0));

    const txInput = tx.inputs[0] as TxIn;
    const txOutBn = txOutBnMap.get(txInput.inputTxId, txInput.inputTxNOut);
    const execScript = txOutBn?.txOut.script as Script;
    const sigBuf = txInput.script.chunks[0]?.buf as WebBuf;
    expect(sigBuf?.length).toBe(65);
    const pubKeyBuf = txInput.script.chunks[1]?.buf as WebBuf;
    expect(pubKeyBuf?.length).toBe(33);

    const stack = [sigBuf, pubKeyBuf];
    const hashCache = new HashCache();

    const scriptInterpreter = ScriptInterpreter.fromOutputScriptTx(
      execScript,
      tx,
      new U32BE(0),
      stack,
      new U64BE(100n),
      hashCache,
    );

    const result = scriptInterpreter.evalScript();
    expect(!!result.value).toBe(true);
  });

  test("should sign two inputs", () => {
    const key = KeyPair.fromRandom();
    const pkh = Pkh.fromPubKeyBuf(key.pubKey.toBuf());
    const script = Script.fromPkhOutput(pkh);
    const output = new TxOut(new U64BE(100), script);
    txBuilder.addOutput(output);
    txBuilder.addOutput(output);

    const tx = txBuilder.build();

    expect(tx.inputs.length).toBe(2);
    expect(tx.outputs.length).toBe(2);
    expect(tx.outputs[0]?.value.bn).toEqual(BigInt(100));
    expect(tx.outputs[1]?.value.bn).toEqual(BigInt(100));

    txSigner = new TxSigner(tx, txOutBnMap, pkhKeyMap, new U32BE(0n));
    const signed1 = txSigner.sign(new U32BE(0));
    expect(signed1).toBeTruthy();
    const signed2 = txSigner.sign(new U32BE(1));
    expect(signed2).toBeTruthy();

    const txInput1 = tx.inputs[0] as TxIn;
    const txOutput1 = txOutBnMap.get(txInput1.inputTxId, txInput1.inputTxNOut);
    const execScript1 = txOutput1?.txOut.script as Script;
    const sigBuf1 = txInput1.script.chunks[0]?.buf as WebBuf;
    expect(sigBuf1?.length).toBe(65);
    const pubKeyBuf1 = txInput1.script.chunks[1]?.buf as WebBuf;
    expect(pubKeyBuf1?.length).toBe(33);

    const stack1 = [sigBuf1, pubKeyBuf1];
    const hashCache1 = new HashCache();

    const scriptInterpreter1 = ScriptInterpreter.fromOutputScriptTx(
      execScript1,
      tx,
      new U32BE(0),
      stack1,
      new U64BE(100n),
      hashCache1,
    );

    const result1 = scriptInterpreter1.evalScript();
    expect(!!result1.value).toBe(true);

    const txInput2 = tx.inputs[1] as TxIn;
    const txOutput2 = txOutBnMap.get(txInput2.inputTxId, txInput2.inputTxNOut);
    const execScript2 = txOutput2?.txOut.script as Script;
    const sigBuf2 = txInput2.script.chunks[0]?.buf as WebBuf;
    expect(sigBuf2?.length).toBe(65);
    const pubKeyBuf2 = txInput2.script.chunks[1]?.buf as WebBuf;
    expect(pubKeyBuf2?.length).toBe(33);

    const stack2 = [sigBuf2, pubKeyBuf2];
    const hashCache2 = new HashCache();

    const scriptInterpreter2 = ScriptInterpreter.fromOutputScriptTx(
      execScript2,
      tx,
      new U32BE(1),
      stack2,
      new U64BE(100n),
      hashCache2,
    );

    const result2 = scriptInterpreter2.evalScript();
    expect(!!result2.value).toBe(true);
  });
});

File: rs/earthbucks_lib/tests/tx_signer_test.rs
use earthbucks_lib::key_pair::KeyPair;
use earthbucks_lib::pkh::Pkh;
use earthbucks_lib::pkh_key_map::PkhKeyMap;
use earthbucks_lib::pub_key::PubKey;
use earthbucks_lib::script::Script;
use earthbucks_lib::script_interpreter::ScriptInterpreter;
use earthbucks_lib::tx::HashCache;
use earthbucks_lib::tx_builder::TxBuilder;
use earthbucks_lib::tx_out::TxOut;
use earthbucks_lib::tx_out_bn_map::TxOutBnMap;
use earthbucks_lib::tx_signature::*;
use earthbucks_lib::tx_signer::*;

#[test]
fn should_sign_a_tx() {
    let mut tx_out_bn_map = TxOutBnMap::new();
    let mut pkh_key_map = PkhKeyMap::new();

    // generate 5 keys, 5 outputs, and add them to the txOutMap
    for i in 0..5 {
        let key = KeyPair::from_random();
        let pkh = Pkh::from_pub_key_buffer(key.pub_key.buf.to_vec());
        pkh_key_map.add(key.clone(), &pkh.buf.clone());
        let script = Script::from_pkh_output(&pkh.buf.clone());
        let output = TxOut::new(100, script);
        let block_num = 0;
        tx_out_bn_map.add(vec![0; 32].as_slice(), i, output, block_num);
    }

    let mut tx_builder = TxBuilder::new(&tx_out_bn_map, Script::from_empty(), 0);
    let tx_out = TxOut::new(50, Script::from_empty());
    tx_builder.add_output(tx_out);

    let tx = tx_builder.build().unwrap();

    assert_eq!(tx.inputs.len(), 1);
    assert_eq!(tx.outputs.len(), 2);
    assert_eq!(tx.outputs[0].value, 50);

    let mut tx_signer = TxSigner::new(tx.clone(), &tx_out_bn_map, &pkh_key_map, 0);
    let tx_res = tx_signer.sign_input(0);
    let signed_tx = tx_signer.tx;
    assert!(tx_res.is_ok());

    let tx_input = &signed_tx.inputs[0];
    let tx_out_bn = tx_out_bn_map
        .get(&tx_input.input_tx_id.clone(), tx_input.input_tx_out_num)
        .unwrap();
    let exec_script = tx_out_bn.tx_out.script.clone();
    let sig_buf = tx_input.script.chunks[0].buffer.clone().unwrap();
    assert_eq!(sig_buf.len(), TxSignature::SIZE);
    let pub_key_buf = tx_input.script.chunks[1].buffer.clone().unwrap();
    assert_eq!(pub_key_buf.len(), PubKey::SIZE);

    let stack = vec![sig_buf, pub_key_buf];

    let mut hash_cache = HashCache::new();
    let mut script_interpreter = ScriptInterpreter::from_output_script_tx(
        exec_script,
        signed_tx,
        0,
        stack,
        100,
        &mut hash_cache,
    );

    let result = script_interpreter.eval_script();
    assert!(result);
}

#[test]
fn should_sign_two_inputs() {
    let mut tx_out_bn_map = TxOutBnMap::new();
    let mut pkh_key_map = PkhKeyMap::new();

    // generate 5 keys, 5 outputs, and add them to the txOutMap
    for i in 0..5 {
        let key = KeyPair::from_random();
        let pkh = Pkh::from_pub_key_buffer(key.pub_key.buf.to_vec());
        pkh_key_map.add(key.clone(), &pkh.buf.clone());
        let script = Script::from_pkh_output(&pkh.buf.clone());
        let output = TxOut::new(100, script);
        let block_num = 0;
        tx_out_bn_map.add(vec![0; 32].as_slice(), i, output, block_num);
    }

    let mut tx_builder = TxBuilder::new(&tx_out_bn_map, Script::from_empty(), 0);
    let tx_out = TxOut::new(100, Script::from_empty());
    tx_builder.add_output(tx_out.clone());
    tx_builder.add_output(tx_out.clone());

    let tx = tx_builder.build().unwrap();

    assert_eq!(tx.inputs.len(), 2);
    assert_eq!(tx.outputs.len(), 2);
    assert_eq!(tx.outputs[0].value, 100);
    assert_eq!(tx.outputs[1].value, 100);

    let mut tx_signer = TxSigner::new(tx.clone(), &tx_out_bn_map, &pkh_key_map, 0);
    let tx_res1 = tx_signer.sign_input(0);
    let tx_res2 = tx_signer.sign_input(1);
    let signed_tx = tx_signer.tx;
    assert!(tx_res1.is_ok());
    assert!(tx_res2.is_ok());

    let tx_input_1 = &signed_tx.inputs[0];
    let tx_out_bn_1 = tx_out_bn_map
        .get(&tx_input_1.input_tx_id.clone(), tx_input_1.input_tx_out_num)
        .unwrap();
    let tx_out_1 = tx_out_bn_1.tx_out.clone();
    let exec_script_1 = tx_out_1.script.clone();
    let sig_buf_1 = tx_input_1.script.chunks[0].buffer.clone().unwrap();
    assert_eq!(sig_buf_1.len(), TxSignature::SIZE);
    let pub_key_buf_1 = tx_input_1.script.chunks[1].buffer.clone().unwrap();
    assert_eq!(pub_key_buf_1.len(), PubKey::SIZE);

    let stack_1 = vec![sig_buf_1, pub_key_buf_1];

    let mut hash_cache = HashCache::new();
    let mut script_interpreter_1 = ScriptInterpreter::from_output_script_tx(
        exec_script_1,
        signed_tx.clone(),
        0,
        stack_1,
        100,
        &mut hash_cache,
    );

    let result_1 = script_interpreter_1.eval_script();
    assert!(result_1);

    let tx_input_2 = &signed_tx.inputs[1];
    let tx_out_bn_2 = tx_out_bn_map
        .get(&tx_input_2.input_tx_id.clone(), tx_input_2.input_tx_out_num)
        .unwrap();
    let tx_out_2 = tx_out_bn_2.tx_out.clone();
    let exec_script_2 = tx_out_2.script.clone();
    let sig_buf_2 = tx_input_2.script.chunks[0].buffer.clone().unwrap();
    assert_eq!(sig_buf_2.len(), TxSignature::SIZE);
    let pub_key_buf_2 = tx_input_2.script.chunks[1].buffer.clone().unwrap();
    assert_eq!(pub_key_buf_2.len(), PubKey::SIZE);

    let stack_2 = vec![sig_buf_2, pub_key_buf_2];
    let mut hash_cache = HashCache::new();

    let mut script_interpreter_2 = ScriptInterpreter::from_output_script_tx(
        exec_script_2,
        signed_tx.clone(),
        1,
        stack_2,
        100,
        &mut hash_cache,
    );

    let result_2 = script_interpreter_2.eval_script();
    assert!(result_2);
}
